In [1]:
#import libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df = pd.read_csv('balance-scale.data', sep=',')
df.head()

B  1  1.1  1.2  1.3
0  R  1    1    1    2
1  R  1    1    1    3
2  R  1    1    1    4
3  R  1    1    1    5
4  R  1    1    2    1

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 624 entries, 0 to 623
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   B       624 non-null    object
 1   1       624 non-null    int64 
 2   1.1     624 non-null    int64 
 3   1.2     624 non-null    int64 
 4   1.3     624 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 24.5+ KB


In [4]:
#Column B is the class Name which we will be predicting

In [7]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
X = df.drop('B', axis=1)
y = df[['B']]

In [8]:
#split our data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [10]:
#fit decision tree model using critirion "gini"
clf_model = DecisionTreeClassifier(criterion='gini', random_state=42, max_depth=3, min_samples_leaf=5)
clf_model.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=3, min_samples_leaf=5, random_state=42)

In [11]:
#make predictions
y_pred = clf_model.predict(X_test)

In [12]:
#check metrics

from sklearn.metrics import confusion_matrix, classification_report
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[ 0  7  5]
 [ 0 60 27]
 [ 0 15 74]]
              precision    recall  f1-score   support

           B       0.00      0.00      0.00        12
           L       0.73      0.69      0.71        87
           R       0.70      0.83      0.76        89

    accuracy                           0.71       188
   macro avg       0.48      0.51      0.49       188
weighted avg       0.67      0.71      0.69       188



/usr/local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [13]:
#create a target and see how decision tree can be plotted
target = list(df['B'].unique())
feature_names = list(X.columns)

In [17]:
#plot tree using export text function
from sklearn.tree import export_text
r = export_text(clf_model, feature_names=feature_names)
print(r)

|--- 1 <= 2.50
|   |--- 1.3 <= 2.50
|   |   |--- 1.2 <= 2.50
|   |   |   |--- class: L
|   |   |--- 1.2 >  2.50
|   |   |   |--- class: R
|   |--- 1.3 >  2.50
|   |   |--- 1.2 <= 1.50
|   |   |   |--- class: R
|   |   |--- 1.2 >  1.50
|   |   |   |--- class: R
|--- 1 >  2.50
|   |--- 1.1 <= 2.50
|   |   |--- 1.2 <= 1.50
|   |   |   |--- class: L
|   |   |--- 1.2 >  1.50
|   |   |   |--- class: R
|   |--- 1.1 >  2.50
|   |   |--- 1.3 <= 3.50
|   |   |   |--- class: L
|   |   |--- 1.3 >  3.50
|   |   |   |--- class: L



'1' is Left weight, and it has been split into two points: 1.3 'right distance' and 1.2 'left distance

Reading from top to bottom we can see how step by step the tree is classifiying our variables
example from top
if left weight is less than or equal to 2.5, split to right distance, if right distance is <=2.5 split to left distance, if left distance is <=2.5, classify as class L